### PR TITLE
[FLOC-3039] Enable profiling

### PR DESCRIPTION
**Pull request by adamtheturtle**
Fixes https://clusterhq.atlassian.net/browse/FLOC-3039

### CODE REVIEW

File: flocker/control/script.py [hunk 1]
```diff
@@ -77,6 +81,35 @@ def main(self, reactor, options):
...
+    def output_profile(signal, frame):
+        """
+        Dump profiling statistics to a file.
+        This disables profiling.
+
+        :param int signal: See ``signal.signal``.
+        :param frame: None or frame object. See ``signal.signal``.
+        """
+        current_time = time.strftime("%Y%m%d%H%M%S")
+        path = FilePath('/etc/flocker/profile-{}'.format(current_time))
```
Comment (jongiddy) [inline flocker/control/script.py]: `/etc` is the wrong place to put this file.  Either `/var/...` or `/tmp` would be better.

Comment (adamtheturtle) [inline flocker/control/script.py]: This is something we weren't sure about. Do you have a preference for an exact path?

Comment (jongiddy) [inline flocker/control/script.py]: Well, `/var/flocker` would be good, but requires that you ensure that directory is created and has correct permissions. So, `/tmp` may be easier, or `/var/tmp` if you need it to survive a reboot.

File: docs/using/administering/debugging.rst [hunk 2]
```diff
@@ -119,6 +119,47 @@ Alternatively, the information can be gathered manually using the following comm
...
+
+See :py:mod:`pstats` for details on how to extract information from this file.
+For example:
+
+.. code-block:: python
+
+   import pstats
+
+   profile = pstats.Stats('profile-20150917161214')
+   profile.print_stats()
```
Comment (jongiddy) [inline docs/using/administering/debugging.rst]: Use a more useful line, like (copied from pstats page):
```python
profile.sort_stats('cumulative').print_stats(10)
```

File: docs/using/administering/debugging.rst [hunk 3]
```diff
@@ -119,6 +119,47 @@ Alternatively, the information can be gathered manually using the following comm
...
+Profiling data will then be collected until a signal to dump the profiling data is received.
+To dump the current profile run the following command as root on the control node:
+
+.. prompt:: bash #
+
+   pkill -SIGUSR2 flocker-control
+
+This will output the profiling data to a file named :file:`/etc/flocker/profile-<TIMESTAMP>`.
+This will also disable profiling.
+When profiling is re-enabled, new profiling data will be added to the existing profiling data until the profile is dumped again.
```
Comment (jongiddy) [inline docs/using/administering/debugging.rst]: This line is a bit confusing. Do you mean something like: "If profiling is re-enabled in the same run of the control server, new profiling data will be added to all previously collected profiling data when the profile is dumped again"

File: docs/using/administering/debugging.rst [hunk 4]
```diff
@@ -119,6 +119,47 @@ Alternatively, the information can be gathered manually using the following comm
...
+
+Profiling data will then be collected until a signal to dump the profiling data is received.
+To dump the current profile run the following command as root on the control node:
+
+.. prompt:: bash #
+
+   pkill -SIGUSR2 flocker-control
+
+This will output the profiling data to a file named :file:`/etc/flocker/profile-<TIMESTAMP>`.
+This will also disable profiling.
```
Comment (jongiddy) [inline docs/using/administering/debugging.rst]: :question: Change this to emphasise first that SIGUSR1 starts profiling and SIGUSR2 stops profiling. Then mention that when profiling is stopped, all profiling data collected up to that point, including from previous intervals of profiling, will be dumped. This clearly matches the two signals as delimiting phases of profiling, while also explaining how to get the profiling data.

File: flocker/control/script.py [hunk 5]
```diff
@@ -77,6 +81,35 @@ def main(self, reactor, options):
...
+    def enable_profiling(signal, frame):
+        """
+        Enable profiling of the control service.
+
+        :param int signal: See ``signal.signal``.
+        :param frame: None or frame object. See ``signal.signal``.
+        """
+        pr.enable()
+
+    def output_profile(signal, frame):
```
Comment (jongiddy) [inline flocker/control/script.py]: :question: As for the docs, consider renaming to `disable_profiling` to emphasise on/off semantics, and document profile output as a side effect.

**Comment by jongiddy** [comment]
Tested with a control service on CentOS 7. Signal handling works to start and stop profiling.

Please address the comments, and resubmit.

File: flocker/control/script.py (hunk 1 above)
Comment (adamtheturtle) [inline flocker/control/script.py]: I'll go with `/tmp` unless there are any objections.

File: docs/using/administering/debugging.rst (hunk 3 above)
Comment (adamtheturtle) [inline docs/using/administering/debugging.rst]: Yes, I will change this.

File: flocker/control/script.py (hunk 5 above)
Comment (adamtheturtle) [inline flocker/control/script.py]: :+1:

File: flocker/control/script.py [hunk 6]
```diff
@@ -77,6 +81,35 @@ def main(self, reactor, options):
 
 
 def flocker_control_main():
+    pr = cProfile.Profile()
```
Comment (jongiddy) [inline flocker/control/script.py]: Bonus comment: Use `Profile(time.clock)` to get cpu-time instead of wallclock time.

**Comment by adamtheturtle** [comment]
All points addressed but the script couldn't seem to write to `/tmp` or `/var/tmp` (I didn't investigate why) so I used `var/lib/flocker`.

**Comment by jongiddy** [comment]
Tested on Ubuntu 14.04 this time round. Works fine.

Please merge when all-green.

**Pull request merged**